design/tls-client-verification.md

```diff
@@ -0,0 +1,156 @@
+# Client certificate validation (mTLS)
+
+Status: Draft
+
+## Goals
+
+- Allow client certificate validation (mTLS) on contour (performed by Envoy)
+- Allow various ways for client certificate validation; Spki, Hash and CA
+- Document mTLS configuration
+
+## Non Goals
+
+- Support for k8s ingress documents.
+- Repeat configuration details described in Envoy documentation (make references)
+
+## Background
+
+In TLS (and https) only the server is authenticated with a
+certificate, for instance you as a client can be sure that you speak
+with your bank and not some malicious site. But sometimes also the client
+must be authenticated. As noted in
+[wikipedia](https://en.wikipedia.org/wiki/Mutual_authentication)
+client certification (mTLS) is not very common for end-users but is
+more widespread for business-to-business (B2B) applications (which
+may use gRPC and REST APIs).
+
+I can't give a real-life example but it is easy to imagine cases where
```

Comment:
Here are some common example scenarios for protecting REST APIs:

* Elasticsearch did not support authentication, [until very recently](https://www.elastic.co/blog/tips-to-secure-elasticsearch-clusters-for-free-with-encryption-users-and-more). The common pattern was to put reverse proxy in front of Elasticsearch to handle TLS and authentication
* Prometheus does not support either [TLS](https://prometheus.io/docs/guides/tls-encryption/) or [authentication](https://prometheus.io/docs/guides/basic-auth/). Common patter is to put reverse proxy in front.

Reply:
Thanks!